src/io/sb3/toSb3.ts

```diff
@@ -1,163 +1,1120 @@
+    // stored inline with the input they correspond to, not as separate blocks
+    // with IDs.
+    //
+    // Within this code, we use the primimtiveOrOpCode option to determine how
+    // the shadow should be serialized. If it is a number, it's referring to a
+    // "primitive", the term uses for shadows when they are in the compressed
+    // form. If it's a string, it is a (shadow) block opcode, and should be
+    // serialized in the expanded form.
+
+    const { blockData, getBroadcastId, parentId, primitiveOrOpCode } = options;
+
+    let shadowValue = null;
+
+    if (primitiveOrOpCode === BIS.BROADCAST_PRIMITIVE) {
+      // Broadcast primitives, unlike all other primitives, expect two values:
+      // the broadcast name and its ID.
+      shadowValue = [BIS.BROADCAST_PRIMITIVE, value, getBroadcastId(value)];
+    } else if (primitiveOrOpCode === BIS.COLOR_PICKER_PRIMITIVE) {
+      // Color primitive. Convert the {r, g, b} object into hex form.
+      const hex = (k: string): string => (value || { r: 0, g: 0, b: 0 })[k].toString(16).padStart(2, "0");
+      shadowValue = [BIS.COLOR_PICKER_PRIMITIVE, "#" + hex("r") + hex("g") + hex("b")];
+    } else if (typeof primitiveOrOpCode === "number") {
+      // Primitive shadow, can be stored in compressed form.
+      shadowValue = [primitiveOrOpCode, value];
+    } else {
+      // Note: Only 1-field shadow blocks are supported.
+      const shadowOpCode = primitiveOrOpCode;
+      const fieldEntries = sb3.fieldTypeMap[shadowOpCode];
+      if (fieldEntries) {
+        const fieldKey = Object.keys(fieldEntries)[0];
+        const fields = { [fieldKey]: [value] };
 
-function extractCostumes(target: Target): sb3.Costume[] {
-  return target.costumes.map(costume => {
-    const sb3Costume: sb3.Costume = {
-      assetId: costume.md5,
-      name: costume.name,
-      md5ext: `${costume.md5}.${costume.ext}`,
-      dataFormat: costume.ext,
-      rotationCenterX: costume.centerX / 2,
-      rotationCenterY: costume.centerY / 2
+        const id = generateId();
+
+        blockData[id] = {
+          opcode: shadowOpCode,
+
+          next: null,
+          parent: parentId,
+
+          fields,
+          inputs: {},
+
+          shadow: true,
+          topLevel: false
+        };
+
+        shadowValue = id;
+      }
+    }
+
+    return { shadowValue };
+  }
+
+  interface SerializeInputsToInputsOptions<PassedInputs extends { [key: string]: BlockInput.Any }> {
+    stage: Stage;
+    target: Target;
+
+    blockData: sb3.Target["blocks"];
+
+    getBroadcastId: GetBroadcastId;
+    customBlockDataMap: CustomBlockDataMap;
+
+    block: Block;
+    inputEntries;
+
+    initialValues: {
+      [key in keyof PassedInputs]: any;
     };
-    return sb3Costume;
-  });
-}
+  }
+
+  function serializeInputsToInputs<PassedInputs extends { [key: string]: BlockInput.Any }>(
+    inputs: PassedInputs,
+    options: SerializeInputsToInputsOptions<PassedInputs>
```

Comment:
Is there any info in `inputEntries` that doesn't exist in `inputs`?

Reply:
Yep - this reminds me `inputEntries` doesn't currently have a type definition. I should definitely give it one. Essentially, it refers to a value in `inputPrimitiveOrShadowMap` (see interfaces.ts). It describes how the input shadow should be serialized (i.e. as which kind of primitive or which menu opcode), which is data that sb-edit doesn't provide elsewhere.